relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -62,23 +69,42 @@ fn get_trace_sampling_rule<'a>(
     check_unsupported_rules(processing_enabled, sampling_config)?;
 
     let rule = or_ok_none!(sampling_config.get_matching_trace_rule(sampling_context, ip_addr));
-    Ok(Some((rule, sampling_context.trace_id)))
+    let sample_rate = match sampling_config.mode {
+        SamplingMode::Received => rule.sample_rate,
+        SamplingMode::Total => sampling_context.adjusted_sample_rate(rule.sample_rate),
+    };
+
+    Ok(Some(SamplingSpec {
+        sample_rate,
+        rule_id: rule.id,
+        seed: sampling_context.trace_id,
+    }))
 }
 
-fn get_event_sampling_rule<'a>(
+fn get_event_sampling_rule(
     processing_enabled: bool,
-    project_state: &'a ProjectState,
+    project_state: &ProjectState,
+    sampling_context: Option<&DynamicSamplingContext>,
     event: Option<&Event>,
     ip_addr: Option<IpAddr>,
-) -> Result<Option<(&'a SamplingRule, Uuid)>, SamplingResult> {
+) -> Result<Option<SamplingSpec>, SamplingResult> {
     let event = or_ok_none!(event);
     let event_id = or_ok_none!(event.id.value());
 
     let sampling_config = or_ok_none!(&project_state.config.dynamic_sampling);
     check_unsupported_rules(processing_enabled, sampling_config)?;
 
     let rule = or_ok_none!(sampling_config.get_matching_event_rule(event, ip_addr));
-    Ok(Some((rule, event_id.0)))
+    let sample_rate = match (sampling_context, sampling_config.mode) {
+        (Some(ctx), SamplingMode::Total) => ctx.adjusted_sample_rate(rule.sample_rate),
+        _ => rule.sample_rate,
+    };
+
+    Ok(Some(SamplingSpec {
+        sample_rate,
+        rule_id: rule.id,
+        seed: event_id.0,
+    }))
```

Comment:
Can we add tests for these functions? I would even volunteer to write some :)

Reply:
You read my mind - was already in the works. See [5081542](https://github.com/getsentry/relay/pull/1591/commits/5081542700579e0267beec56bd01ce77acb25193)